cirq/study/flatten_expressions.py

```diff
@@ -0,0 +1,368 @@
+    of `flatten` as a shortcut.
+
+    Args:
+        val: The value to copy with substituted parameters.
+
+    Example:
+        >>> qubit = cirq.LineQubit(0)
+        >>> a = sympy.Symbol('a')
+        >>> circuit = cirq.Circuit.from_ops(
+        ...     cirq.X(qubit) ** (a/4),
+        ...     cirq.Y(qubit) ** (1-a/2),
+        ... )
+        >>> print(circuit)
+        0: ───X^(a/4)───Y^(1 - a/2)───
+
+        >>> sweep = cirq.Linspace(a, start=0, stop=3, length=4)
+        >>> list(sweep)  #doctest: +NORMALIZE_WHITESPACE
+        [cirq.ParamResolver(OrderedDict([('a', 0.0)])),
+         cirq.ParamResolver(OrderedDict([('a', 1.0)])),
+         cirq.ParamResolver(OrderedDict([('a', 2.0)])),
+         cirq.ParamResolver(OrderedDict([('a', 3.0)]))]
+
+        >>> c_flat, expr_map = cirq.flatten(circuit)
+        >>> print(c_flat)
+        0: ───X^(<a/4>)───Y^(<1 - a/2>)───
+        >>> expr_map
+        cirq.ExpressionMap({a/4: <a/4>, 1 - a/2: <1 - a/2>})
+
+        >>> new_sweep = expr_map.transform_sweep(sweep)
+        >>> list(new_sweep)  #doctest: +NORMALIZE_WHITESPACE
+        [cirq.ParamResolver(OrderedDict([('<a/4>', 0.0), ('<1 - a/2>', 1.0)])),
+         cirq.ParamResolver(OrderedDict([('<a/4>', 0.25), ('<1 - a/2>', 0.5)])),
+         cirq.ParamResolver(OrderedDict([('<a/4>', 0.5), ('<1 - a/2>', 0.0)])),
+         cirq.ParamResolver(OrderedDict([('<a/4>', 0.75), ('<1 - a/2>', -0.5)])\
+)]
+
+        >>> for params in sweep:  # Original
+        ...     print(circuit, '=>', end=' ')
+        ...     print(cirq.resolve_parameters(circuit, params))
+        0: ───X^(a/4)───Y^(1 - a/2)─── => 0: ───X^0───Y───
+        0: ───X^(a/4)───Y^(1 - a/2)─── => 0: ───X^0.25───Y^0.5───
+        0: ───X^(a/4)───Y^(1 - a/2)─── => 0: ───X^0.5───Y^0───
+        0: ───X^(a/4)───Y^(1 - a/2)─── => 0: ───X^0.75───Y^-0.5───
+        >>> for params in new_sweep:  # Flattened
+        ...     print(c_flat, '=>', end=' ')
+        ...     print(cirq.resolve_parameters(c_flat, params))
+        0: ───X^(<a/4>)───Y^(<1 - a/2>)─── => 0: ───X^0───Y───
+        0: ───X^(<a/4>)───Y^(<1 - a/2>)─── => 0: ───X^0.25───Y^0.5───
+        0: ───X^(<a/4>)───Y^(<1 - a/2>)─── => 0: ───X^0.5───Y^0───
+        0: ───X^(<a/4>)───Y^(<1 - a/2>)─── => 0: ───X^0.75───Y^-0.5───
+
+    Returns:
+        A tuple containing the value with flattened parameters and the
+        expression map mapping expressions in `val` to symbols.
+    """
+    flattener = _ParamFlattener()
+    val_flat = flattener.flatten(val)
+    expr_map = ExpressionMap(flattener.param_dict)
+    return val_flat, expr_map
+
+
+def flatten_with_sweep(val: Any,
+                       sweep: Union[sweeps.Sweep, List[resolver.ParamResolver]]
+                      ) -> Tuple[Any, sweeps.Sweep]:
+    """Creates a copy of `val` with any symbols or expressions replaced with
+    new symbols.  `val` can be a `Circuit`, `Gate`, `Operation`, or other
+    type.  Also creates a transformed `Sweep` that resolves the new value.
```

Comment:
Same comment here and below.  I would recommend more descriptive text to this and the following function.

Reply:
Done.